crates/lib/src/action/worker.rs

```diff
@@ -32,17 +32,132 @@ pub async fn worker(
 	events_tx: priority::Sender<Event, Priority>,
 	events: priority::Receiver<Event, Priority>,
 ) -> Result<(), CriticalError> {
-	let mut last = Instant::now();
-	let mut set = Vec::new();
-	let process = ProcessHolder::default();
-	let outcome_gen = OutcomeWorker::newgen();
+	let mut processes: HashMap<SupervisorId, ProcessData> = HashMap::new();
+
+	// Insert the base process from the initial `WorkingData`.
+	let base_process_id = SupervisorId::default();
+	processes.insert(
+		base_process_id,
+		ProcessData {
+			working: working.clone(),
+			process: ProcessHolder::default(),
+			outcome_gen: OutcomeWorker::newgen(),
+		},
+	);
+
+	while let Some(mut set) = throttle_collect(
+		events.clone(),
+		working.clone(),
+		errors.clone(),
+		Instant::now(),
+	)
+	.await?
+	{
+		#[allow(clippy::iter_with_drain)]
+		let events = Arc::from(set.drain(..).collect::<Vec<_>>().into_boxed_slice());
+		let action = Action::new(
+			Arc::clone(&events),
+			Arc::from(
+				processes
+					.keys()
+					.copied()
+					.collect::<Vec<_>>()
+					.into_boxed_slice(),
+			),
+		);
+		debug!("running action handler");
+		let action_handler = {
+			let wrk = working.borrow();
+			wrk.action_handler.clone()
+		};
 
-	loop {
-		if events.is_closed() {
-			trace!("events channel closed, stopping");
-			break;
+		let outcomes = action.outcomes.clone();
+		let err = action_handler
+			.call(action)
+			.await
+			.map_err(|e| rte("action worker", e.as_ref()));
+		if let Err(err) = err {
+			errors.send(err).await?;
+			debug!("action handler errored, skipping");
+			continue;
 		}
 
+		for (pid, (resolution, event_set)) in outcomes.lock().await.iter() {
+			let found = match resolution {
+				Resolution::Apply(outcome) => processes
+					.get(pid)
+					.map(|data| (outcome.clone(), data.clone())),
+				Resolution::Start(cmd) => {
+					assert!(processes.get(pid).is_none());
+					let mut wrk = working.borrow().clone();
+					wrk.commands = vec![cmd.clone()];
+
+					let (_, wrk_rx) = watch::channel(wrk);
```

Comment:
that is a good point! 

maybe, detaching the command that is executed by the `Process` from the `WorkingData` would allow the reuse of the same `WorkingData` for all the `Process`s?

Reply:
Is anything from the WorkingData apart from the command being used? If not, then there's no need to have workingdata passed down anymore.

Reply:
hmm, i believe so, though i may be wrong...

it looks like `OutcomeWorker` still gets the `post_spawn_handler` and `pre_spawn_handler` from the `WorkingData`?

Reply:
Ah yes, so it is quite important that those do get updated. If you can reuse the workingdata across all processes as you said, that would be grand.

Reply:
okie, yea, that makes sense!

it looks like, and definitely correct me if i'm missing something else, the only place working is then used from here is when the `Outcome` is matched in the `OutcomeWorker` here:

```rust
(_, Outcome::Start) => {
  let (cmds, grouped, pre_spawn_handler, post_spawn_handler) = {
    let wrk = self.working.borrow();
      (
      wrk.commands.clone(),
      wrk.grouped,
      wrk.pre_spawn_handler.clone(),
      wrk.post_spawn_handler.clone(),
      )
    };
    
    if cmds.is_empty() {
      warn!("tried to start commands without anything to run");
    } else {
      trace!("spawning supervisor for command");
      let sup = Supervisor::new(RequiredArgs {
      errors: self.errors_c.clone(),
      events: self.events_c.clone(),
      commands: cmds,
      supervisor_id: self.supervisor_id,
      grouped,
      actioned_events: self.events.clone(),
      pre_spawn_handler,
      post_spawn_handler,
    })?;
    notry!(self.process.replace(sup));
  }
}
```

So, maybe the commands could be passed to the `OutcomeWorker` separately from the `WorkingData`, as they are really just used in this step. That would, of course, add another argument to the `OutcomeWorker::spawn()` method, which already needs to use `#[allow(too_many_arguments)]`, so this may not be ideal... what do you think? would something like this:

```rust
pub fn spawn(
  outcome: Outcome,
  events: Arc<[Event]>,
  // new parameter
  cmds: Vec<Command>,
  working: Receiver<WorkingData>,
  process: ProcessHolder,
  supervisor_id: SupervisorId,
  gencheck: Arc<AtomicUsize>,
  errors_c: mpsc::Sender<RuntimeError>,
  events_c: priority::Sender<Event, Priority>,
) {
```

be a decent solution?

Reply:
Yeah, looks good.